### ISSUE TITLE
Memory leak in ListView when deleting an item from the ItemsSource

### ISSUE THREAD

**Issue by MichaelRumpler** [xamarin/Xamarin.Forms#3396]
### Description

If you bind a ListView.ItemsSource to an ObservableCollection and then remove an item from that collection, then the associated renderers' Dispose method is not called. Only when the page is popped, then iOS disposes them, but Android and UWP do not. I did not check the other platforms.

### Steps to Reproduce

1. Create a ViewModel with an ObservableCollection<T>
2. Bind a ListView.ItemsSource to that collection
3. Delete an item from the ObservableCollection<T>

### Expected Behavior

The renderers of all elements in the associated cell are disposed.

### Actual Behavior

The Dispose method of those renderers is never called. I have to assume, that they are still in memory.

### Basic Information

- Version with issue: tested with XF 3.0 up to 3.2-pre1

### Reproduction Link

In the attached repro project I created a CustomLabel with a CustomLabelRenderer just to see when which methods are called. These classes only log, when the important methods are called and with which arguments, but otherwise don't change behavior.
Both the SplashPage and the MainPage show the log of those method calls.
[Repro_MemoryLeak.zip](https://github.com/xamarin/Xamarin.Forms/files/2227658/Repro_MemoryLeak.zip)

**Comment by andreinitescu** [comment]
Does calling `GC.Collect(); GC.WaitForPendingFinalizers(); ` manually makes any difference? I'm not suggesting this is a solution, but only to try see if it's really a leak.

**Comment by MichaelRumpler** [comment]
No, they are still not disposed.

**Comment by andreinitescu** [comment]
I could not see any GC. calls in the repro you attached. Is it there or did you just tested it?

**Comment by MichaelRumpler** [comment]
It's not in the attached project.
I already tested GC.Collect before I submitted this bug. After your comment I also added WaitForPendingFinalizers. I triggered them with an additional button on the SplashPage after the MainPage was popped, but no difference.

**Comment by jassmith** [comment]
iOS doesn't dispose the renderers until the ListView goes away. Are you expecting the cache to get cleared?  New renderers also shouldn't be created if you were to then add a new item into the ListView after removing the first assuming its using the same template.

**Comment by MichaelRumpler** [comment]
No, if it is for caching then the behavior on iOS is ok. But the others should also dispose the renderers which are not needed anymore.

**Comment by andreinitescu** [comment]
@jassmith I see you tagged this with needs-info, what kind of info is needed?

**Comment by andreinitescu** [comment]
@MichaelRumpler Are you aware of any info which is needed?

**Comment by MichaelRumpler** [comment]
The only question he asked is "Are you expecting the cache to get cleared?" and I answered it.

On bugzilla they expected the people who opened a bug to change the status back - and almost no one knew that so the bugs were all stalled. I hope they don't expect that here too because I don't know how to remove that "needs-info" label. I don't think that I'm allowed to.

**Comment by andreinitescu** [comment]
@MichaelRumpler We don't have the permission to add or remove labels on issues. The label was set almost 2 weeks ago for apparently no reason.

**Comment by samhouts** [comment]
Not to worry! This issue is not forgotten. To be sure I'm understanding the issue here:

1. Renderers in a ListView are not disposed when the ListView is disposed on Android and UWP when the ItemSource is an ObservableCollection.
2. iOS is not affected by this bug. The renderers are properly disposed when the ListView is disposed.
3. If the ItemSource is _not_ an ObservableCollection, the renderers are properly disposed on all platforms when the ListView is disposed.

Is that accurate?

I'll leave the needs-info label on until confirmation of the above information. Thanks!

**Comment by hartez** [comment]
Thanks for the repro!

> Android bound to ObservableCollection<string>: only visible renderers are disposed 👎

When you say "only visible renderers", what do you mean? How are you making that determination?

**Comment by MichaelRumpler** [comment]
When you delete an item from the ObservableCollection, then the respective view is removed from the UI. If I cannot see it in the UI then it is not visible.

**Comment by hartez** [comment]
@MichaelRumpler So "visible" as in "remaining" - gotcha. I was interpreting "visible" as "not offscreen" (i.e., not scrolled out of view). Thanks for clarifying.
